### PR TITLE
Implement replay and state control function for os provision probe

### PR DESCRIPTION
**Pull request by hu-weihua**
This pull request covers below items:
*  Implement replay function, customer can offer a time point, ``osdeploy`` tool will search history log, and find out the last state of target node
*  Implement state control, using state machine
* Add max waiting time for monitor 
*  refine some code logic

Below are some real output of  ``osdeploy``, 

1 do monitor with verbose information

    # xcatprobe osdeploy -n c910f02c04p04 -V
    All pre_defined nodes are valid                                                                                   [ OK ]
    The installation network interface is eth0                                                                        [INFO]
    -------------------------------------------------------------
    Start to capture every message during os provision process......
    -------------------------------------------------------------

    [c910f02c04p04] Node status is changed to powering-off
    [c910f02c04p04] Node status is changed to powering-off
    [c910f02c04p04] Receive BOOTREQUEST from 42:6a:be:22:e7:03 via eth0
    [c910f02c04p04] Send BOOTREPLY on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
    [c910f02c04p04] Receive BOOTREQUEST from 42:6a:be:22:e7:03 via eth0
    [c910f02c04p04] Send BOOTREPLY on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
    [c910f02c04p04] Via TFTP download /boot/grub2/grub2-c910f02c04p04
    ..............
    [c910f02c04p04] Via TFTP download /boot/grub2/powerpc-ieee1275/crypto.mod
    [c910f02c04p04] Node status is changed to powering-on
    [c910f02c04p04] Via TFTP download /boot/grub2/powerpc-ieee1275/gettext.mod
    .............
    [c910f02c04p04] Via TFTP download /xcat/osimage/rhels7.1-ppc64-install-compute/vmlinuz
    [c910f02c04p04] Via TFTP download /xcat/osimage/rhels7.1-ppc64-install-compute/initrd.img
    [c910f02c04p04] Receive DHCPDISCOVER via eth0
    [c910f02c04p04] Send DHCPOFFER on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
    [c910f02c04p04] Receive DHCPREQUEST from 42:6a:be:22:e7:03 for 10.2.4.4 via eth0
    [c910f02c04p04] Send DHCPACK on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
    [c910f02c04p04] Via HTTP get /install/autoinst/c910f02c04p04
    .............
    [c910f02c04p04] Via HTTP get /install/rhels7.1/ppc64/images/product.img
    [c910f02c04p04] Send DHCPACK on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
    [c910f02c04p04] Node status is changed to installing
    [c910f02c04p04] Via HTTP get /install/rhels7.1/ppc64/.treeinfo
    .............
    [c910f02c04p04] Via HTTP get /tftpboot/mypostscripts/mypostscript.c910f02c04p04
    [c910f02c04p04] xcat:  Tue Aug  9 02:27:48 EDT 2016 Running postscript: syslog
    .............
    [c910f02c04p04] xcat:  Tue Aug  9 02:28:07 EDT 2016 postscript setbootfromdisk return with 0
    [c910f02c04p04] Node status is changed to booting
    [c910f02c04p04] Receive DHCPDISCOVER via eth0
    [c910f02c04p04] Send DHCPOFFER on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
    [c910f02c04p04] Receive DHCPREQUEST from 42:6a:be:22:e7:03 for 10.2.4.4 via eth0
    [c910f02c04p04] Send DHCPACK on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
    [c910f02c04p04] xcat:  running /xcatpost/mypostscript.post
    [c910f02c04p04] xcatpostinit1:  Tue Aug  9 02:31:47 EDT 2016 Running postscript: otherpkgs
    [c910f02c04p04] xcatpostinit1: Tue Aug  9 02:31:47 EDT 2016 Running postscript: otherpkgs
    [c910f02c04p04] xcat:  Tue Aug  9 02:31:47 EDT 2016 Running postscript: otherpkgs
    [c910f02c04p04] xcat:  Running otherpkgs
    [c910f02c04p04] xcatpostinit1:  ./otherpkgs: no extra rpms to install
    [c910f02c04p04] xcatpostinit1: ./otherpkgs: no extra rpms to install
    [c910f02c04p04] xcatpostinit1:  Postscript: otherpkgs exited with code 0
    [c910f02c04p04] xcatpostinit1: Postscript: otherpkgs exited with code 0
    [c910f02c04p04] xcat:  Tue Aug  9 02:31:47 EDT 2016 postscript otherpkgs return with 0
    [c910f02c04p04] xcat:  ready
    [c910f02c04p04] xcat:  done
    [c910f02c04p04] Node status is changed to booted
    All nodes need to monitor have finished os provision process                                                      [ OK ]

    =============================================================
    =  The summary of os provision:
    =============================================================

    [c910f02c04p04]
            Node status is changed to powering-off
            Node status is changed to powering-off
            Receive BOOTREQUEST from 42:6a:be:22:e7:03 via eth0
            Send BOOTREPLY on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
            Receive BOOTREQUEST from 42:6a:be:22:e7:03 via eth0
            Send BOOTREPLY on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
            Via TFTP download /boot/grub2/grub2-c910f02c04p04
            ..............
            Via TFTP download /boot/grub2/powerpc-ieee1275/crypto.mod
            Node status is changed to powering-on
            Via TFTP download /boot/grub2/powerpc-ieee1275/gettext.mod
            .............
            Via TFTP download /xcat/osimage/rhels7.1-ppc64-install-compute/initrd.img
            Receive DHCPDISCOVER via eth0
            Send DHCPOFFER on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
            Receive DHCPREQUEST from 42:6a:be:22:e7:03 for 10.2.4.4 via eth0
            Send DHCPACK on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
            Via HTTP get /install/autoinst/c910f02c04p04
            Via HTTP ..........
            Via HTTP get /install/rhels7.1/ppc64/images/updates.img
            Via HTTP get /install/rhels7.1/ppc64/images/product.img
            Send DHCPACK on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
            Node status is changed to installing
            Via HTTP get /install/rhels7.1/ppc64/.treeinfo
            Via HTTP ..........
            Via HTTP get /install/postscripts/repos/rhels7.1/ppc64/?C=D;O=D
            Via HTTP get /tftpboot/mypostscripts/mypostscript.c910f02c04p04
            xcat:  Tue Aug  9 02:27:48 EDT 2016 Running postscript: syslog
            ...............
            xcat:  Tue Aug  9 02:28:07 EDT 2016 postscript setbootfromdisk return with 0
            Node status is changed to booting
            Receive DHCPDISCOVER via eth0
            Send DHCPOFFER on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
            Receive DHCPREQUEST from 42:6a:be:22:e7:03 for 10.2.4.4 via eth0
            Send DHCPACK on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
            xcat:  running /xcatpost/mypostscript.post
            xcatpostinit1:  Tue Aug  9 02:31:47 EDT 2016 Running postscript: otherpkgs
            xcatpostinit1: Tue Aug  9 02:31:47 EDT 2016 Running postscript: otherpkgs
            xcat:  Tue Aug  9 02:31:47 EDT 2016 Running postscript: otherpkgs
            xcat:  Running otherpkgs
            xcatpostinit1:  ./otherpkgs: no extra rpms to install
            xcatpostinit1: ./otherpkgs: no extra rpms to install
            xcatpostinit1:  Postscript: otherpkgs exited with code 0
            xcatpostinit1: Postscript: otherpkgs exited with code 0
            xcat:  Tue Aug  9 02:31:47 EDT 2016 postscript otherpkgs return with 0
            xcat:  ready
            xcat:  done
            Node status is changed to booted
    [c910f02c04p04] deployment completed                                                                              [ OK ]

2 do replay with verbose information

    # xcatprobe osdeploy -n c910f02c04p04,c910f02c04p05 -r 24 -V
    All pre_defined nodes are valid                                                                                   [ OK ]
    The installation network interface is eth0                                                                        [INFO]
    Start to search logs after 'Mon Aug  8 02:37:56 2016', please waiting for a while.............

    =============================================================
    =  The summary of os provision:
    =============================================================

    [c910f02c04p04]
            Node status is changed to powering-off
            Node status is changed to powering-on
            Send DHCPACK on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
            Node status is changed to powering-off
            Node status is changed to powering-off
            Receive BOOTREQUEST from 42:6a:be:22:e7:03 via eth0
            Send BOOTREPLY on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
            Receive BOOTREQUEST from 42:6a:be:22:e7:03 via eth0
            Send BOOTREPLY on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
            Via TFTP download /boot/grub2/grub2-c910f02c04p04
            Via TFTP download /boot/grub2/powerpc-ieee1275/normal.mod
            Via TFTP download /boot/grub2/powerpc-ieee1275/terminal.mod
            Node status is changed to powering-on
            Via TFTP download /boot/grub2/powerpc-ieee1275/crypto.mod
            ...........
            Via TFTP download /xcat/osimage/rhels7.1-ppc64-install-compute/initrd.img
            Receive DHCPDISCOVER via eth0
            Send DHCPOFFER on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
            Receive DHCPREQUEST from 42:6a:be:22:e7:03 for 10.2.4.4 via eth0
            Send DHCPACK on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
            Via HTTP get /install/autoinst/c910f02c04p04
            Via HTTP ..........
            Via HTTP get /install/rhels7.1/ppc64/images/updates.img
            Via HTTP get /install/rhels7.1/ppc64/images/product.img
            Send DHCPACK on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
            Node status is changed to installing
            Via HTTP get /install/rhels7.1/ppc64/.treeinfo
            Via HTTP ..........
            Via HTTP get /install/postscripts/_xcat/?C=S;O=A
            Via HTTP get /install/postscripts/_xcat/?C=D;O=A
            xcat:  Tue Aug  9 02:27:48 EDT 2016 Running postscript: syslog
            .............
            xcat:  Tue Aug  9 02:28:07 EDT 2016 postscript setbootfromdisk return with 0
            Node status is changed to booting
            Receive DHCPDISCOVER via eth0
            Send DHCPOFFER on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
            Receive DHCPREQUEST from 42:6a:be:22:e7:03 for 10.2.4.4 via eth0
            Send DHCPACK on 10.2.4.4 back to 42:6a:be:22:e7:03 via eth0
            xcat:  running /xcatpost/mypostscript.post
            xcat:  Tue Aug  9 02:31:47 EDT 2016 Running postscript: otherpkgs
            xcat:  Running otherpkgs
            xcat:  Tue Aug  9 02:31:47 EDT 2016 postscript otherpkgs return with 0
            xcat:  ready
            xcat:  done
            Node status is changed to booted
            xcat:  /xcatpost/mypostscript.post return
            xcatpostinit1:  Tue Aug  9 02:31:47 EDT 2016 Running postscript: otherpkgs
            xcatpostinit1:  ./otherpkgs: no extra rpms to install
            xcatpostinit1:  Postscript: otherpkgs exited with code 0
            xcat.provision:  provision completed.(c910f02c04p04)
            xcatpostinit1: Tue Aug  9 02:31:47 EDT 2016 Running postscript: otherpkgs
            xcatpostinit1: ./otherpkgs: no extra rpms to install
            xcatpostinit1: Postscript: otherpkgs exited with code 0
    [c910f02c04p04] deployment completed                                                                              [ OK ]
    [c910f02c04p05]
            Node status is changed to powering-on
            Receive DHCPDISCOVER via eth0
            Send DHCPOFFER on 10.2.4.5 back to 42:6a:b4:ce:72:03 via eth0
            Receive DHCPREQUEST from 42:6a:b4:ce:72:03 for 10.2.4.5 via eth0
            Send DHCPACK on 10.2.4.5 back to 42:6a:b4:ce:72:03 via eth0
            Send DHCPACK on 10.2.4.5 back to 42:6a:b4:ce:72:03 via eth0
            Send DHCPACK on 10.2.4.5 back to 42:6a:b4:ce:72:03 via eth0
            Send DHCPACK on 10.2.4.5 back to 42:6a:b4:ce:72:03 via eth0
    [c910f02c04p05] deployment failed                                                                                 [FAIL]
            There are more than one possible process satisfy c910f02c04p05 situation
            possible process "deploy", pass sever_reboot stage, something wrong during loaded_kernel_and_initrd stage
            possible process "reboot", pass sever_reboot stage, something wrong during run_postbootscript stage
            possible process "reboot", pass sever_reboot stage, something wrong during done stage

3 do replay without verbose information
 
    # xcatprobe osdeploy -n c910f02c04p04,c910f02c04p05 -r 24
    All pre_defined nodes are valid                                                                                   [ OK ]
    The installation network interface is eth0                                                                        [INFO]
    Start to search logs after 'Mon Aug  8 02:38:35 2016', please waiting for a while.............

    =============================================================
    =  The summary of os provision:
    =============================================================

    [c910f02c04p04] deployment completed                                                                              [ OK ]
    [c910f02c04p05] deployment failed                                                                                 [FAIL]
            There are more than one possible process satisfy c910f02c04p05 situation
            possible process "deploy", pass sever_reboot stage, something wrong during loaded_kernel_and_initrd stage
            possible process "reboot", pass sever_reboot stage, something wrong during run_postbootscript stage
            possible process "reboot", pass sever_reboot stage, something wrong during done stage

### CODE REVIEW

File: xCAT-probe/subcmds/osdeploy [hunk 1]
```diff
@@ -54,58 +177,58 @@ Options:
...
+            next;
+        }
+        $monitor_nodes{$node}{"rst"} = 0;
+        my @macs = split(/\|/, $nodecheckrst{$node}{"mac"});
+        foreach my $mac (@macs) {
+            if ($mac =~ /\!\*NOIP\*/) {
+                $mac =~ s/\!\*NOIP\*//g;
+                $macmap{$mac}{"ip"}   = "NOIP";
+                $macmap{$mac}{"node"} = $node;
+            } else {
```
Comment (zet809) [inline xCAT-probe/subcmds/osdeploy]: The mac attribute can be like this: 
c910f03c25: mac=98:be:94:00:c4:78|98:be:94:00:c4:79!c910f03c25-enP4p1s0f1
The string c910f03c25-enP4p1s0f1 behind of "!" can be resolved to an IP address.
But for you case, the $mac include it which is not correct.

File: xCAT-probe/subcmds/osdeploy [hunk 2]
```diff
@@ -9,35 +9,158 @@ use probe_utils;
...
 
+#- %monitor_nodes structure-------
+# $monitor_nodes{nodename}{"status"}  #useless now
+# $monitor_nodes{nodename}{"rst"}
+#---------------------------------
+
+# provision state machine
+my %state_set = (
+    "unknown"                  => 0,
+    "sever_reboot"             => 1,
```
Comment (zet809) [inline xCAT-probe/subcmds/osdeploy]: sever_reboot ==> server_reboot

File: xCAT-probe/subcmds/osdeploy [hunk 3]
```diff
@@ -407,30 +545,113 @@ sub dump_history {
     print "$title\n";
```
Comment (zet809) [inline xCAT-probe/subcmds/osdeploy]: Pls use the output msg function consistent.

File: xCAT-probe/subcmds/osdeploy [hunk 4]
```diff
@@ -612,13 +812,187 @@ Start to capture every message during os provision process......
...
+                #TODO  handle compression files
+            }
+
+            #print "--->load $file\n";
+            my $line;
+            my $historynum = 0;
+            last unless ($line = <$fd>);
+            chomp($line);
+            my $needrotate = 0;
+            my $istorynum  = 0;
```
Comment (zet809) [inline xCAT-probe/subcmds/osdeploy]: Never use.

**Comment by hu-weihua** [comment]
@zet809 , I have modified the code, Could you help to review and merge. thanks

**Comment by zet809** [comment]
@hu-weihua Thx wei hua, I will merge this pull request and have opened defect #1692 to trace the improvement of the code logic for the replying when doing deploy probe.

**Pull request merged**